### PR TITLE
Add wrapping selection when word is wrapped

### LINKED ISSUE #4009: Make double-click at the edge select across the edge
# Environment

Windows build number: 19041.1
Windows Terminal version (if applicable): 0.7.3451.0

# Steps to reproduce

- You may need to configure `wordDelimiters` to equal `" \\()\"',;<>~!@#$^*|+[]{}~│"`
- Open Windows Terminal at default window width, using Ubuntu
- Run `echo 'really long string of text, really particularly long indeed, typing typing typing typing, also this url: https://example.com'`
- Double click on the URL

# Expected behavior
 

The entirety of the URL should be selected.

# Actual behavior

Depending on whether you double click on the URL before or after the line wrap, either just `https://example` or `.com` is selected.

### PR DESCRIPTION
**Pull request by js324**
## Summary of the Pull Request
Added wrapping to highlighted selection when selecting a word, added tests for it
## References and Relevant Issues
#4009 
## Detailed Description of the Pull Request / Additional comments
- Modified GetWordStart and GetWordEnd and their helpers to no longer be bounded by the right and left viewport ranges
- Kept same functionality (does not wrap) when selecting wrapped whitespace
- Added tests to TextBufferTests.cpp to include cases of wrapping text

Just a thought, we currently select the "Line" on triple-click, but would it make more sense to include the wrapping functionality to this as well? 

## Validation Steps Performed
- Ran locally and verified selection works properly
- Tests passed locally
## PR Checklist
- [x] Closes #4009 
- [x] Tests added/passed
- [ ] Documentation updated
   - If checked, please file a pull request on [our docs repo](https://github.com/MicrosoftDocs/terminal) and link it here: #xxx
- [ ] Schema updated (if necessary)

### CODE REVIEW

**Comment by DHowett** [comment]
HECK YES!

Thank you for taking this on. I'm sorry we've been quiet on the reviews -- we've had some in-Windows commitments to land. I'll make sure we get some eyes on this in the soontimes.

File: src/buffer/out/textBuffer.cpp [hunk 1]
```diff
@@ -1247,36 +1247,32 @@ til::point TextBuffer::_GetWordStartForAccessibility(const til::point target, co
...
         }
+        bufferSize.DecrementInBounds(result);
     }
 
-    // move off of delimiter and onto word start
-    if (!stayAtOrigin && _GetDelimiterClassAt(result, wordDelimiters) != DelimiterClass::RegularChar)
-    {
-        bufferSize.IncrementInBounds(result);
-    }
+    // move off of delimeter
```
Comment (zadjii-msft) [inline src/buffer/out/textBuffer.cpp:1274]:
```suggestion
    // move off of delimiter
```

**Review by zadjii-msft — CHANGES_REQUESTED** [review]
Sorry this took so long to get back to!

I'm mostly blocking as a nit over the spellbot updates.

~I'm also super curious about the ControlChar thing - I wouldn't expect that to affect this code really at all, but maybe I'm mistaken? Could you elaborate why that particular delimiter class matters in this case?~

OH it's not about _control characters_. It's specifically about **spaces**. Okay now I get it.

File: src/buffer/out/textBuffer.cpp [hunk 2]
```diff
@@ -1294,10 +1290,16 @@ til::point TextBuffer::_GetWordStartForSelection(const til::point target, const
     const auto bufferSize = GetSize();
 
     const auto initialDelimiter = _GetDelimiterClassAt(result, wordDelimiters);
+    const bool isControlChar = initialDelimiter == DelimiterClass::ControlChar;
```
[the remark this answers is not in the document]
Comment (zadjii-msft) [reply src/buffer/out/textBuffer.cpp:1293]: Figured it out as I was submitting the review 🤦

File: .github/actions/spelling/expect/04cdb9b77d6827c0202f51acd4205b017015bfff.txt
[the remark this answers is not in the document]
Comment (js324) [reply .github/actions/spelling/expect/04cdb9b77d6827c0202f51acd4205b017015bfff.txt:1]: Agreed on the typos but some of the words are related to my comments in TextBufferTests.cpp where I'm writing what the buffer should look like in the new tests. Should I keep those or just remove the comment?

**Comment by js324** [comment]
Could I get some help how to resolve this new spelling-bot report? It doesn't seem related to my changes and I'm not sure how to resolve it cleanly 😅

File: src/buffer/out/textBuffer.cpp [hunk 3]
```diff
@@ -1375,25 +1377,21 @@ til::point TextBuffer::_GetWordEndForAccessibility(const til::point target, cons
...
-            ++iter;
+            bufferSize.IncrementInBounds(result);
         }
 
-        while (iter && iter.Pos() != limit && _GetDelimiterClassAt(iter.Pos(), wordDelimiters) != DelimiterClass::RegularChar)
+        while (result != limit && result != bufferSize.BottomRightInclusive() && _GetDelimiterClassAt(result, wordDelimiters) != DelimiterClass::RegularChar)
         {
             // expand to the beginning of the NEXT word
-            ++iter;
+            bufferSize.IncrementInBounds(result);
```
Comment (lhecker) [inline src/buffer/out/textBuffer.cpp:1389]: I've been meaning to review this, but I still can't get to this. I'm still hung-up on fixing all remaining bugs that I can find before we ship the next big version of this project with Windows.

From skimming it, I believe that this PR is mostly correct, but I'm somewhat worried that it goes into the wrong direction. In particular, there's this line. This one doesn't properly work with wide glyphs (= 2 column wide glyphs like Chinese/Japanese), whereas the previous code probably did it correctly from what I can tell. If you get a chance, I'd suggest testing this PR with such characters.

To be clear, we'll definitely review this PR in time for the next version of Windows Terminal, that is planned to first ship in Preview sometime early next year, but it's possible that a proper review may be delayed until then.

**Comment by js324** [comment]
Really sorry about that, was trying to clean up the commits while merging the latest changes in main and I accidentally closed this. I submitted a new PR

**Pull request closed without merging**